### ISSUE TITLE
Automatically typed Literal strings in tuples in lists cause invariance errors

### ISSUE THREAD

**Issue by Vegemash** [microsoft/pyright#5366]
Possibly related to  #5321

**Describe the bug**
When the type of a string, within a tuple, within a list is automatically determined, it is regarded as a Literal.
list[tuple[Literal["whatever"]]]
This can then not be assigned to a list[tuple[str]] 

**To Reproduce**
run pyright 1.1.315 on the code below

**Expected behavior**
No type errors, you should be able to assign a literal string to a string

**Code or Screenshots**

```python
from typing import Literal

A = [("LITERAL",)]
B: list[tuple[str]] = A # Error here

C = ("LITERAL",)
D: tuple[str] = C

E = [Literal["LITERAL"]]
F: list[str] = E # Error here

G = ["LITERAL"]
F: list[str] = G

```

**VS Code extension or command-line**
command line tool

**Comment by erictraut** [comment]
This is by design, not a bug. Please refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference) for details about pyright's default type inference behaviors for list and tuple expressions. Regardless of which default type inference behaviors a type checker adopts, there will always be cases where it is necessary to override these defaults. To do so, you can provide explicit type declarations.

```python
A: list[tuple[str]] = [("LITERAL",)]
B: list[tuple[str]] = A
```

**Comment by benhoyt** [comment]
@erictraut I'm running into this issue, but it seems strange it's "by design", because it's only started happening since Pyright 1.1.315 (and 1.1.316 is the same). Versions 1.1.314 and all the versions I've tried before that (.225, .250, .300, .310-.313) don't have a problem with it. Why did the behaviour suddenly change, at least without a significant notice in the release notes -- but neither the .315 or .316 release notes seem to mention a "Behaviour Change" related to this?

I see https://github.com/microsoft/pyright/issues/5321 and its fix in https://github.com/microsoft/pyright/pull/5323, and that seems related (though there are no tuples involved, and they seem to be necessary here), but there's no more commentary there. So is #5321 saying that this is a bug that was there all along? Regardless, the fact that this fails seems very unintuitive to me, and to others I've asked.

Here's a cut-down version of the specific manifestation of the issue I had. The first two examples fail type checking, and the last three examples succeed:

```python
def f(x: list[tuple[str, int]]) -> None:
    print(x)

# Example 1 - fails
arg1 = [('a', 1)]
reveal_type(arg1)
f(arg1)
"""
/path/to/src.py:7:3 - error: Argument of type "list[tuple[Literal['a'], Literal[1]]]" cannot be assigned to parameter "x" of type "list[tuple[str, int]]" in function "f"
  "list[tuple[Literal['a'], Literal[1]]]" is incompatible with "list[tuple[str, int]]"
    TypeVar "_T@list" is invariant
      "tuple[Literal['a'], Literal[1]]" is incompatible with "tuple[str, int]"
        Tuple entry 1 is incorrect type
          "Literal['a']" is incompatible with "str" (reportGeneralTypeIssues)
"""

# Example 2 - fails
arg2 = [(str('a'), 1)]
reveal_type(arg2)
f(arg2)
"""
/path/to/src.py:12:3 - error: Argument of type "list[tuple[str, Literal[1]]]" cannot be assigned to parameter "x" of type "list[tuple[str, int]]" in function "f"
  "list[tuple[str, Literal[1]]]" is incompatible with "list[tuple[str, int]]"
    TypeVar "_T@list" is invariant
      "tuple[str, Literal[1]]" is incompatible with "tuple[str, int]"
        Tuple entry 2 is incorrect type
          "Literal[1]" is incompatible with "int" (reportGeneralTypeIssues)
"""

# Example 3 - succeeds
arg3 = [(str('a'), int(1))]
reveal_type(arg3)
f(arg3)

# Example 4 -- succeeds
arg4: list[tuple[str, int]] = [('a', 1)]
reveal_type(arg4)
f(arg4)

# Example 5 -- succeeds
f([('a', 1)])
```

I've read through the docs that you linked to, and from those rules I can see why it's being inferred as `list[tuple[Literal["LITERAL"]]]`, but I'm unsure why that type allowed to fulfil the less specific type `list[tuple[str]]`, just as `Literal["LITERAL"]` can fulfil type `str`. The message even says `"Literal['a']" is incompatible with "str"` -- but why is that? Isn't `str` a superset of `Literal['a']`? The ["literal types" PEP](https://peps.python.org/pep-0586/) says "For example, Literal[3] is a subtype of int."

And in the following small example (without list), Pyright (1.1.316) succeeds fine, even though we're trying to pass a tuple with literals to a function that takes a `tuple[str, int]`:

```python
def g(x: tuple[str, int]) -> None:
    print(x)

x = ('a', 1)
reveal_type(x)          # tuple[Literal['a'], Literal[1]]
g(x)                    # succeeds
y: tuple[str, int] = x  # succeeds
```

So if you can spare a few minutes, I'd love a clearer explanation of why this doesn't work. To a long-time Python programmer who's relatively new to type annotations, this seems very unintuitive to me. I realize Pyright tries to be more clever than MyPy about type inference, but MyPy doesn't have this "problem", so maybe Pyright's being too clever here, and as such, not matching people's intuition?

**Comment by hmc-cs-mdrissi** [comment]
List is invariant while tuple is covariant. So function that wants list[tuple[X]] will be stricter then function that wants tuple[X] in regard to sub typing.

If your function doesn’t need to mutate list it is generally better to use Sequence/Iterable as input type instead. Covariant generics are easier to work with. If you use invariant generics then at times you may need to have specific annotation for list. Doing mypy strategy is alternative but either choice produces some false positives. Whether tuple of strings should be literals or str has no defined answer. https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types is a good explanation on generic variance.

**Comment by erictraut** [comment]
Apologies for the slow response. I'm on holiday, and my access to the internet is intermittent. @hmc-cs-mdrissi already answered your question, but I'll provide some additional details.

Pyright 1.1.315 fixed a serious type checking bug, which explains the difference in behavior that you're seeing here.

The type parameter for `list` is invariant, which means that the types must match exactly for compatibility. While `int` is type compatible with `float`, `list[int]` is not type compatible with `list[float]`. Likewise, `Literal[0]` is type compatible with `int`, but `list[Literal[0]]` is not type compatible with `list[int]`. Versions of pyright prior to 1.1.315 allowed the latter without flagging it as a type violation.

You can work around your issue in a couple of ways.

First, you can use an explicit type annotation to override the default inference rules for tuples and lists.

```python
def f(x: list[tuple[str, int]]) -> None:
    print(x)

arg1: list[tuple[str, int]] = [('a', 1)]
f(arg1)
```

Second, you can use immutable forms of container classes that use covariant (rather than invariant) type parameters. The invariant form of `list` is `Sequence`.

```python
def f(x: Sequence[tuple[str, int]]) -> None:
    print(x)

arg1 = [('a', 1)]
f(arg1)
```

**Comment by benhoyt** [comment]
Thanks @erictraut and @hmc-cs-mdrissi -- those are very helpful responses, and a good explanation of what invariant vs covariant means. I still find this less than intuitive, but at least it's clear why now. Yeah, I think the thing to do in my case is change the signature to `Sequence[T]` rather than `list[T]`.

**Comment by benhoyt** [comment]
@erictraut BTW, in your last sentence did you mean "The *covariant* form of `list` is `Sequence`"?

**Comment by erictraut** [comment]
Yes, I meant _covariant_. Apologies for any confusion.
